presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java

```diff
@@ -204,25 +204,33 @@ public PlanNode visitTableFinish(TableFinishNode node, RewriteContext<FragmentPr
     @Override
     public PlanNode visitSequence(SequenceNode node, RewriteContext<FragmentProperties> context)
     {
-        // Since this is topologically sorted by the LogicalCtePlanner, need to make sure that execution order follows
-        // Can be optimized further to avoid non dependents from getting blocked
-        int cteProducerCount = node.getCteProducers().size();
-        checkArgument(cteProducerCount >= 1, "Sequence Node has 0 CTE producers");
-        PlanNode source = node.getCteProducers().get(cteProducerCount - 1);
-        FragmentProperties childProperties = new FragmentProperties(new PartitioningScheme(
-                Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()),
-                source.getOutputVariables()));
-        SubPlan lastSubPlan = buildSubPlan(source, childProperties, context);
-
-        for (int sourceIndex = cteProducerCount - 2; sourceIndex >= 0; sourceIndex--) {
-            source = node.getCteProducers().get(sourceIndex);
-            childProperties = new FragmentProperties(new PartitioningScheme(
+        // To ensure that the execution order is maintained, we use an independent dependency graph.
+        // This approach creates subgraphs sequentially, enhancing control over the execution flow. However, there are optimization opportunities:
+        // 1. Can consider blocking only the CTEConsumer stages that are in a reading state.
+        //    This approach sounds good on paper may not be ideal as it can block the entire query, leading to resource wastage since no progress can be made until the writing operations are complete.
+        // 2. ToDo: Another improvement will be to schedule the execution of subgraphs based on their order in the overall execution plan instead of a topological sorting done here
```

Comment:
general question again, but still trying to understand the basics. can you draw out what a fragmented plan with dependent ctes would look like before and after this change?

Reply:
Sure, lets take an example case of **Graph(a, b)**  or sql 
`WITH t as (SELECT * FROM orders), t1 as (Select * FROM customer) SELECT * FROM t JOIN t1 ON true`

Lets say that the Plan would be the below for simplicity.

```
...
     Exchange
        Sequence:
                           cteProducer1
                           cteProducer2
       PrimarySource:
                      Exchange
                                  Join 
                                     ....

```

Lets say the primary source creates subplan (subPrimary1->subPrimary2) 
and creates a list of subplans for the cteProducers as (subCteProducerT1 and subCteProducerT2)
The code is a bit tricker here but in the visitSequence we add the the cteproducers to the context created by the parent exchange. ([here](https://github.com/prestodb/presto/blob/1367d3caa79a3a9b62e3821f0cb27760a0d39e4c/presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java#L300) is actually where the parent exchange creates new properties or context)

so before the subplans look like


```
SubParent -> subPrimary1 -> subPrimary2
          -> subCteProducer1 -> subCteProducer2
```
[code](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java#L222) is where we were adding the subCteProducer2 as a child of subCteProducer1

Now we would add the subCteProducer to the same level as the subCteProducer1

```
SubParent ->  subPrimary1 -> subPrimary2
           -> subCteProducer1 
           -> subCteProducer2
```
                                     
In both cases similar structure would be represented in the sectionedPlan